### ISSUE TITLE
Provider does not create resources in the order of the Kustomize output

### ISSUE THREAD

**Issue by toteshreyas** [kbst/terraform-provider-kustomize#36]
Hi,
I want to integrate the IAM roles on the pod level using kustomization provider, kindly let me know how can I do that. Is there any way so that we can make our pod secure.

**Comment by pst** [comment]
I'm afraid it's not clear to me what you are trying to do. Can you show an example of what you have so far?

**Comment by toteshreyas** [comment]
Hi Philipp,
1. Defined the 7 resources in kustomization.yaml and tried to run it via terraform kustomization provider but it is not taking the sequence while executing the resources. Ideally, it tooks that sequence when we manually run using kubectl.
2. while deploying yaml files using Terraform kustomization provider, not able to pass variables in yaml files as we used to do in Terraform. is there any way to pass the variables.
Note:- we are working to install the istio-1.5.2 with custom IAM roles using Terraform.

**Comment by pst** [comment]
Number 2 is a limitation of the provider currently. There is an open issue with discussion on the topic #18 

Regarding number 1, how are you using the provider? What does your HCL look like and what error are you getting?

**Comment by toteshreyas** [comment]
My Kustomization file as below,
 
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: istio-system
resources: 
 - test1.yaml
 - test2.yaml
 - test3.yaml
 - test4.yaml
 - test5.yaml
 - test6.yaml
 - test7.yaml

and main.tf as below

data "kustomization" "test" {
  path = "istio"
}

resource "kustomization_resource" "test" {
  for_each = data.kustomization.test.ids

  manifest = data.kustomization.test.manifests[each.value]
}

where istio is the name of directory, in which I have kept the test 1 to test7 files.

**Comment by toteshreyas** [comment]
Hi, Thanks for your reply. Regarding the issue 1 kindly find below details.

*main.tf as below, istio is the name of the directory in which I have kept all the resources.

data "kustomization" "test" {
  path = "istio"
}
resource "kustomization_resource" "test" { for_each = data.kustomization.test.ids
 manifest = data.kustomization.test.manifests[each.value]
}

*Provider.tf as below.

provider "kustomization" {
 kubeconfig_path = "~/.kube/config"
context = "given my arn here 
}

*kustomization.yaml as below

`` apiVersion: kustomize.config.k8s.io/v1beta1``
``kind: Kustomization``
``namespace: istio-system``
``resources:`` 
 ``- crd.yaml``
 ``- ns.yaml``
 ``- reader.yaml``
 ``- gateway.yaml``
 ``- sidecar.yaml``
 ``- istiod.yaml``

**Comment by pst** [comment]
This looks fine, what's the error you get?

If your question is, how to customize the Istio resources in any way, then you would do that using Kustomize patches. The Terraform provider here replaces Kubectl, because that has benefits regarding purging and more. But it does not currently support making changes to the Kubernetes resources coming from Kustomize easily.

**Comment by toteshreyas** [comment]
I don't want to customize istio resources, the problem is the resource execution in kustomization.yaml file. To install istio correctly I need to run the above files(crd.yaml,ns.yaml etc) which I mentioned in kustomization resource in sequence. The correact sequence is first to install the custom resource definitions, then create namespace and then forth on which I can achieve while run "kubectl apply -k ."  When I used terraform provider for the same, I am getting an issue where kustomization provider is taking its own sequence and installation is failed.
Example error:- 
Error: ResourceCreate: GroupVersionKind 'apiextensions.k8s.io/v1beta1, Kind=CustomResourceDefinition' couldn't find resource (21 retries)

**Comment by pst** [comment]
You have to excuse my confusion, you keep changing what this issue is about it seems.

What version of the provider are you using? This should work and is also verified with a test since `v0.1.0-beta.3`. If this error occurs with the latest version there must be another issue.

**Comment by pst** [comment]
Ok I spent the last 4 hours looking into this more. Using something like this:

```
data "kustomization" "test" {
  path = "test_kustomizations/istio"
}

resource "kustomization_resource" "test" {
  for_each = data.kustomization.test.ids

  manifest = data.kustomization.test.manifests[each.value]
}
```
The ids, are originally in the correct order (the one predefined by Kustomize) which puts namespaces first, then CRDs etc. But, `ids` in the internal Terraform Schema is a set. And that set is [sorted](https://github.com/hashicorp/terraform-plugin-sdk/blob/v1.4.0/helper/schema/set.go#L238) before it's returned.

The whole purpose of the ids field was to keep the order from Kustomize. But as I now learned, that did not work.

Because Terraform doesn't know of any dependencies between the resources coming from Kustomize, the provider waits for a namespace or a CRD to exist before creating a resource. However, Terraform by default does only 10 resources in parallel. So for something like Istio, that has a bigger number of resources, there may be an edge case where all 10 are in the waiting loop until the time out.

Workarounds I found are re-running the apply and/or increasing `--parallelism=50`
